token.go

```diff
@@ -0,0 +1,224 @@
+		return nil, fmt.Errorf("failed to get SigningCertificateV2 from signed attributes: %w", err)
+	}
+	// get candidate signing certificate
+	var candidateSigningCert *x509.Certificate
+	signed := (*cms.ParsedSignedData)(t)
+	if signingCertificateV2.Certificates[0].IssuerSerial.SerialNumber != nil {
+		// use IssuerSerial from SigningCertificateV2 signed attribute
+		if signingCertificateV2.Certificates[0].IssuerSerial.IssuerName.Name.Bytes == nil {
+			return nil, errors.New("issuer name is missing in IssuerSerial of SigningCertificateV2 attribute")
+		}
+		var issuer asn1.RawValue
+		if _, err := asn1.Unmarshal(signingCertificateV2.Certificates[0].IssuerSerial.IssuerName.Name.Bytes, &issuer); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal issuer name: %w", err)
+		}
+		ref := cms.IssuerAndSerialNumber{
+			Issuer:       issuer,
+			SerialNumber: signingCertificateV2.Certificates[0].IssuerSerial.SerialNumber,
+		}
+		candidateSigningCert = signed.GetCertificate(ref)
+	} else {
+		// use sid (unsigned) as IssuerSerial
+		candidateSigningCert = signed.GetCertificate(signerInfo.SignerIdentifier)
+	}
+	if candidateSigningCert == nil {
+		return nil, CertificateNotFoundError(errors.New("signing certificate not found in the timestamp token"))
+	}
+	// validate hash of candidate signing certificate
+	hash := crypto.SHA256 // default hash algorithm is id-sha256
+	var ok bool
+	if signingCertificateV2.Certificates[0].HashAlgorithm.Algorithm != nil {
+		// use hash algorithm from SigningCertificateV2 signed attribute
+		hash, ok = oid.ToHash(signingCertificateV2.Certificates[0].HashAlgorithm.Algorithm)
+		if !ok {
+			return nil, errors.New("unsupported certificate hash algorithm in SigningCertificateV2 attribute")
+		}
+	}
+	certHash, err := hashutil.ComputeHash(hash, candidateSigningCert.Raw)
+	if err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(certHash, signingCertificateV2.Certificates[0].CertHash) {
+		return nil, errors.New("signing certificate hash does not match CertHash in SigningCertificateV2 attribute")
+	}
+	return candidateSigningCert, nil
+}
+
+//	Accuracy ::= SEQUENCE {
+//	 seconds     INTEGER             OPTIONAL,
+//	 millis  [0] INTEGER (1..999)    OPTIONAL,
+//	 micros  [1] INTEGER (1..999)    OPTIONAL }
+type Accuracy struct {
+	Seconds      int `asn1:"optional"`
+	Milliseconds int `asn1:"optional,tag:0"`
+	Microseconds int `asn1:"optional,tag:1"`
+}
+
+//	TSTInfo ::= SEQUENCE {
+//	 version         INTEGER                 { v1(1) },
+//	 policy          TSAPolicyId,
+//	 messageImprint  MessageImprint,
+//	 serialNumber    INTEGER,
+//	 genTime         GeneralizedTime,
+//	 accuracy        Accuracy                OPTIONAL,
+//	 ordering        BOOLEAN                 DEFAULT FALSE,
+//	 nonce           INTEGER                 OPTIONAL,
+//	 tsa             [0] GeneralName         OPTIONAL,
+//	 extensions      [1] IMPLICIT Extensions OPTIONAL }
+type TSTInfo struct {
```

Comment:
we should validate TSTInfo before consuming.

Reply:
@priteshbandi Agree. This is done on the caller side. As we stated in the comment: 
```
// SignedToken returns the timestamp token with signatures.
// Callers should invoke Verify to verify the content before comsumption.
```

Reply:
Relying on caller to invoke verify is not a good design because we are relying on caller to do additional thing when we can we can do right thing for them

Reply:
Code updated.